addon-test-support/@ember/test-helpers/-internal/deprecations.ts

```diff
@@ -0,0 +1,91 @@
+import { BaseContext } from '../setup-context';
+import { registerDeprecationHandler } from '@ember/debug';
+
+export interface DeprecationOptions {
+  id: string;
+  until: string;
+  for?: string;
+  since?: {
+    available: string;
+  };
+}
+
+export interface DeprecationFailure {
+  message: string;
+  options: DeprecationOptions;
+}
+
+const DEPRECATIONS = new WeakMap<BaseContext, Array<DeprecationFailure>>();
+// eslint-disable-next-line require-jsdoc
+export function getDeprecationsForContext(context: BaseContext): Array<DeprecationFailure> {
+  if (!context) {
+    throw new TypeError(
+      `[@ember/test-helpers] could not get deprecations for an invalid test context: '${context}`
+    );
+  }
+
+  let deprecations = DEPRECATIONS.get(context);
+
+  if (!Array.isArray(deprecations)) {
+    deprecations = [];
+    DEPRECATIONS.set(context, deprecations);
+  }
+
+  return deprecations;
+}
+
+// eslint-disable-next-line require-jsdoc
+export function getDeprecationsDuringCallbackForContext(
+  context: BaseContext,
+  callback: CallableFunction
+): Array<DeprecationFailure> | Promise<Array<DeprecationFailure>> {
+  if (!context) {
+    throw new TypeError(
+      `[@ember/test-helpers] could not get deprecations for an invalid test context: '${context}`
+    );
+  }
+
+  const deprecations = getDeprecationsForContext(context);
+  const previousLength = deprecations.length;
+
+  const result = callback();
+
+  if (result !== null && typeof result === 'object' && typeof result.then === 'function') {
```

Comment:
@rwjblue the "async" variant implementation

Reply:
In fact, there's this https://www.npmjs.com/package/is-promise.

Reply:
ember-auto-import is not currently required for consumers, so installing an arbitrary library from NPM (without extra build shenanigans) may end up being a breaking change.